### ISSUE TITLE
PAM+SSL Configuration Failure across iCAT+Resource Server

### ISSUE THREAD

**Issue by jasoncoposky** [irods/irods#2575]
We have iRODS 4.0.3 configured on master and resource servers, with PAM configured on master server. Using the default `ClientServerNegotiationPolicy` as `REFUSE`, everything runs as expected.
 
When enabling `STRICT` `ClientServerNegotiationPolicy` with `irodsClientServerPolicy` `CS_NEG_REQUIRE` option on both master and resource servers, `iinit` fails for `rodsAdmin`
 
On resource server for the PAM based users the `iinit` works, whereas for irods admin user or basic auth scheme user `rods` we get

```
ERROR: [-]      iRODS/lib/core/src/sockComm.cpp:901:connectToRhost :  status [SSL_HANDSHAKE_ERROR]  errno [] -- message []
        [-]     iRODS/lib/core/src/sockComm.cpp:63:sockClientStart :  status [SSL_HANDSHAKE_ERROR]  errno [] -- message [failed to call 'client start']
                [-]     libssl.cpp:699:ssl_client_start :  status [SSL_HANDSHAKE_ERROR]  errno [] -- message [error in SSL_connect | error:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed]
 
ERROR: _rcConnect: connectToRhost error, server on master:1247 is probably down status = -2103000 SSL_HANDSHAKE_ERROR
ERROR: Saved password, but failed to connect to server master
``` 

On master iCAT server, we used signed CA certs:
 
1. `server.key` and `sever.crt` generated from signed CA
2. `cat server.crt >> chain.pem`
3. `openssl dhparam -2 -out dhparams.pem 2048`
 
Updated `irodsctl` file with below SSL parameters:

```
irodsSSLCertificateChainFile=$IRODS_HOME/server/config/chain.pem
export irodsSSLCertificateChainFile
irodsSSLCertificateKeyFile=$IRODS_HOME/server/config/server.key
export irodsSSLCertificateKeyFile
irodsSSLDHParamsFile=$IRODS_HOME/server/config/dhparams.pem
export irodsSSLDHParamsFile
irodsSSLCACertificateFile=$IRODS_HOME/server/config/chain.pem
export irodsSSLCACertificateFile
irodsSSLCACertificatePath=$IRODS_HOME/server/config/
export irodsSSLCACertificatePath
irodsSSLVerifyServer=cert
export irodsSSLVerifyServer
``` 

`.irods/.irodsEnv` of admin as:

```
irodsClientServerNegotiation 'request_server_negotiation'
irodsClientServerPolicy 'CS_NEG_REQUIRE'
irodsEncryptionKeySize '32'
irodsEncryptionSaltSize '8'
irodsEncryptionNumHashRounds '16'
irodsEncryptionAlgorithm 'AES-256-CBC'
irodsDefaultHashScheme 'SHA256'
```
 
On resource server:
1.     Copied the `server.key` `server.crt` `chain.pem` `dhparams.pem` from master
2.     Appended the public certificates internal and root ca to `chain.pem`
3.     Enabled SSL Configs, under `/var/lib/irods/iRODS/config/config.mk`
4.     Updated `irodsctl` file with below SSL parameters

```
irodsSSLCertificateChainFile=$IRODS_HOME/server/config/chain.pem
export irodsSSLCertificateChainFile
irodsSSLCertificateKeyFile=$IRODS_HOME/server/config/server.key
export irodsSSLCertificateKeyFile
irodsSSLDHParamsFile=$IRODS_HOME/server/config/dhparams.pem
export irodsSSLDHParamsFile
irodsSSLCACertificateFile=$IRODS_HOME/server/config/chain.pem
export irodsSSLCACertificateFile
irodsSSLCACertificatePath=$IRODS_HOME/server/config/
export irodsSSLCACertificatePath
irodsSSLVerifyServer=cert
export irodsSSLVerifyServer
```
 
`.irods/.irodsEnv` of admin updated with client server negotiation as:

```
irodsClientServerNegotiation 'request_server_negotiation'
irodsClientServerPolicy 'CS_NEG_REQUIRE'
irodsEncryptionKeySize '32'
irodsEncryptionSaltSize '8'
irodsEncryptionNumHashRounds '16'
irodsEncryptionAlgorithm 'AES-256-CBC'
irodsDefaultHashScheme 'SHA256'
```

OBSERVATIONS

1. Able to authenticate via PAM-SSL using `iinit` for PAM and basic users on master ICAT server
2. Able to authenticate via only PAM on resource server, with `rodsAdmin` as basic authscheme (`iinit` fails)
3. `iput` and `iget`:
  * On resource server - to/from master, to/from resource server works (PAM users only)
  * On master server - to/from master works, to/from resource server (PAM and basic users works)

```
Mar 20 11:34:24 pid:12708 NOTICE: rodsServer Release version rods4.0.3 - API Version d is up
Mar 20 12:23:28 pid:12708 NOTICE: rodsServer caught signal 2, exiting
irodsServer: /tmp/hudson-irodsbuild/workspace/irods-build-automated_release-cen63/external/boost_1_55_0z/boost/thread/pthread/condition_variable_fwd.hpp:81: boost::condition_variable::~condition_variable(): Assertion `!ret' failed.
    LocalHostName:  localhost,  Port Num: 1247.
```

**Comment by trel** [comment]
possibly related SSL/PAM multiple server issue
 - https://groups.google.com/d/msg/irod-chat/SyFwQSfhNC4/PF1jEc8WTaYJ

**Comment by NavyaDabbiru** [comment]
Have created multiple DNS hosts as mentioned in the SSL/PAM multiple server issue using signed certificates.
We noticed the same scenario as before
OBSERVATIONS
1.Able to authenticate via PAM-SSL using  iinit  for PAM and basic users on master ICAT server
2.Able to authenticate via only PAM on resource server, with  rodsAdmin  as basic authscheme ( iinit  fails)

**Comment by JustinKyleJames** [comment]
Tje following describes the testing that was performed to validate this.


SSL Test Using Certs Signed by Certificate Authority

To test this scenario, we created our own certificate authority.  This can be skipped if you have certificates that are signed by a CA.

1. Create root key:
openssl req -x509 -new -nodes -key rootCA.key -days 1024 -out rootCA.pem
2. Create root self signed certification:
openssl req -x509 -new -nodes -key rootCA.key -days 1024 -out rootCA.pem
3. Install this certificate on all three machines – ICAT, ResourceServer1, and CLI:
cp rootCA.pem /etc/ssl/certs
c_rehash /etc/ssl/certs

Steps on ICAT server:

1. Create ICAT key:
openssl genrsa -out ICAT.key 2048
2. Create ICAT cert signing request (.csr):
openssl req -new -key ICAT.key -out ICAT.csr 
3. Sign the cert (.crt):
openssl x509 -req -in ICAT.csr -CA rootCA.pem -CAkey rootCA.key -CAcreateserial -out ICAT.crt -days 500
4. Create chain file:
cat ICAT.crt rootCA.pem > ICAT_chain.pem
5. Set environment variables in ~irods/iRODS/irodsctl:
export irodsSSLDHParamsFile=/var/lib/irods/ssl/dhparams.pem
export irodsSSLCertificateChainFile=/var/lib/irods/ssl/ICAT_chain.pem
export irodsSSLCertificateKeyFile=/var/lib/irods/ssl/ICAT.key
6. Restart server (~/iRODS/irodsctl restart) 
7. Set environment variables for client:
export irodsSSLCACertificateFile=/var/lib/irods/ssl/rootCA.pem   
export irodsSSLVerifyServer=hostname
export irodsSSLCACertificatePath=/etc/ssl/certs
8. Test running iinit and ils

Steps on ResourceServer1 server:

1. Create ResourceServer1 key:
openssl genrsa -out ResourceServer1.key 2048
2. Create ResourceServer1 cert signing request (.csr):
openssl req -new -key ResourceServer1.key -out ResourceServer1.csr 
3. Sign the cert (.crt):
openssl x509 -req -in ResourceServer1.csr -CA rootCA.pem -CAkey rootCA.key -CAcreateserial -out RespirceServer1.crt -days 500
4. Create chain file:
cat ResourceServer1.crt rootCA.pem > ResourceServer1_chain.pem
5. Set environment variables in ~irods/iRODS/irodsctl:
export irodsSSLDHParamsFile=/var/lib/irods/ssl/dhparams.pem
export irodsSSLCertificateChainFile=/var/lib/irods/ssl/ResourceServer1_chain.pem
export irodsSSLCertificateKeyFile=/var/lib/irods/ssl/ResourceServer1.key
6. Restart server (~/iRODS/irodsctl restart) 
7. Set environment variables for client:
export irodsSSLCACertificateFile=/var/lib/irods/ssl/rootCA.pem   
export irodsSSLVerifyServer=hostname
export irodsSSLCACertificatePath=/etc/ssl/certs
8. - Test running iinit and ils

Steps on iCommands Client:


1. Set environment variables for client:
export irodsSSLCACertificateFile=/var/lib/irods/ssl/rootCA.pem
export irodsSSLVerifyServer=hostname
export irodsSSLCACertificatePath=/etc/ssl/certs
2. Test connecting to both ICAT and ResourceServer1(.irodsEnv)
irodsHost ICAT
or
irodsHost ResourceServer1
3. Test running iinit and ils


Test Scenario 1

Tested iRODS connections between CLI, ICAT, and ResourceServer1 using the above configuration and normal authentication. 

Test Scenario 2

Tested the same scenarios with the above configuration and PAM enabled on the CLI and ResourceServer clients.

Test Scenario 3

Removed (unset) the irodsSSLCACertificateFile and irodsSSLACACertificatePath session variables.  The SSL connection succeeded as the code has a default CA certificate path of /etc/ssl/certs/.

Test Scenario 4

To verify the validity of scenario 3, removed the root certificate (rootCA.pem) from /etc/ssl/certs/ on the CLI client.  The SSL connection failed as expected.




Important Note:

1. In 4.0.3, if the root certificate is not in the default location (/etc/ssl/certs/) then  irodsSSLCACertificateFile and irodsSSLCACertificatePath variables must both be set.   This is a defect that is being fixed in 4.1.  In 4.1, only one of these environment variables will need to be set.
